### PR TITLE
feat: fetch options plugin

### PR DESCRIPTION
**Pull request by asiaziola**
solves https://github.com/warp-contracts/warp/issues/245

### CODE REVIEW

**Comment by asiaziola** [comment]
classes where fetch has not been wrapped as it seems to not make much sense (or warp instance is not yet available): `Testing.ts`, `WarpGatewayContractDefinitionLoader.ts`, `WarpGatewayInteractionsLoader`, `smartweave-global.ts`, `ArweaveWrapper.ts`

**Comment by ppedziwiatr** [comment]
> `WarpGatewayContractDefinitionLoader.ts`, `WarpGatewayInteractionsLoader`

imho it would make sense to add a wrapper here

File: src/core/WarpFetchWrapper.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import { LoggerFactory } from '../logging/LoggerFactory';
+import { Warp } from './Warp';
+
+export interface FetchRequest {
+  input: RequestInfo | URL;
+  init: Partial<RequestInit>;
+}
+
+export class WarpFetchWrapper {
+  private readonly logger = LoggerFactory.INST.create('WarpFetchWrapper');
```
Comment (ppedziwiatr) [inline src/core/WarpFetchWrapper.ts:10]: I guess we should start using `WarpFetchWrapper.name` :-)

File: src/core/WarpFetchWrapper.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+  constructor(private warp: Warp) {
+    this.warp = warp;
+  }
+
+  warpFetch(input: RequestInfo | URL, init?: RequestInit): Promise<Response> {
```
Comment (ppedziwiatr) [inline src/core/WarpFetchWrapper.ts:16]: simply `fetch`? the class name already suggests that it is a wrapper...

File: tools/fetch-options-plugin.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import { defaultCacheOptions, WarpFactory } from '../src/core/WarpFactory';
+
+class FetchOptionsPlugin implements WarpPlugin<FetchRequest, RequestInit> {
+  process(request: FetchRequest): Partial<RequestInit> {
+    const url = request.input;
+
+    let fetchOptions = request.init;
+
+    if (url == `https://d1o5nlqr4okus2.cloudfront.net/gateway/sequencer/register`) {
+      fetchOptions = {
```
Comment (ppedziwiatr) [inline tools/fetch-options-plugin.ts:16]: hmm, maybe we should merge the options in the wrapper itself? Imho there's a risk that a plugin developer may forget to do a manual merge...

Comment (asiaziola) [reply tools/fetch-options-plugin.ts:16]: we can, it seemed easier for me to just update the object but maybe you're right.

Comment (ppedziwiatr) [reply tools/fetch-options-plugin.ts:16]: I believe it's more important to make it easier for developers to use than for us to implement ;-)

**Comment by ppedziwiatr** [comment]
hmm, I would rebase, clean-up and merge this into main..hate those long-lasting PRs...

**Comment by ppedziwiatr** [comment]
@asiaziola , how about a rebase and merge? or maybe ask Akord if they've had some time to test the new features...

**Pull request merged**